File: Zadání/pilsprog/1525-trezor/Trezor.java
import java.util.Scanner;
import java.util.StringTokenizer;

/**
 * Reseni finalove ulohy PilsProg - Trezor
 * @author Tomas Potuzak
 */  
public class Trezor {
  /** Scanner pro cteni ze standardniho vstupu */
  private static Scanner sc;
    
  /**
   * Vstupni bod programu
   * @param args parametry prikazove radky (nevyuzite)   
   */     
  public static void main(String[] args) {
    sc = new Scanner(System.in);
    int pocetTrezoru = sc.nextInt();
    sc.nextLine();
    String radka = null;
    JedenTrezor jedenTrezor = null;
        
    for (int i = 0; i < pocetTrezoru; i++) {
      radka = sc.nextLine();
      jedenTrezor = new JedenTrezor(radka);
      System.out.println(jedenTrezor.urciNejmensiPocetOtoceni());
    }
  }
}

/**
 * Trida obsahujici metody a promenne pro reseni jednoho trezoru
 * @author Tomas Potuzak
 */  
class JedenTrezor {
  /** Barva vrcholu pro DFS - bila */
  public static final int BILA = -1;
  /** Barva vrcholu pro DFS - seda */
  public static final int SEDA = -2;
  /** Barva vrcholu pro DFS - cerna */
  public static final int CERNA = -3;
  /** Delka klice k trezoru (pocet cislic) */
  public static final int DELKA_KLICE = 4;
  /** Maximalni mozny pocet otoceni */
  public static final int MAXIMALNI_POCET_OTOCENI = 10000;
  /** Vsechny klice k trezoru */
  private String[] klice;
  /** Prechody mezi klici - matice (neorientovanych) hran grafu, na hlavni diagonale je ulozena barva vrcholu, mimo hlavni diagonalu hodnata BILA znamena, ze tam neni hrana */
  private int[][] prechodyMeziKlici;
  /** Pocet vsech klicu */
  private int pocetKlicu;
  /** Pocet vsech neorientovanych hran (prechodu mezi klici) */
  private int pocetHran;
  /** Pole vsech hran (prechodu mezi klici) */
  private Hrana[] hrany;
  /** Promenna, do ktere DFS zaznamena, ze naslo cestu mezi dvema vrcholy */
  private boolean cestaExistuje;
    
  /**
   * Vytvori novou instanci trezoru ze zadaneho retezce klicu 
   * @param trezorStr retezec obsahujici vsechny klice (s poctem klicu jako prvni cislo)
   */        
  public JedenTrezor(String trezorStr) {
    StringTokenizer st = new StringTokenizer(trezorStr, " ");
    this.pocetKlicu = st.countTokens() - 1;
    
    klice = new String[pocetKlicu];
    st.nextToken(); //Zahozeni prvniho tokenu, obsahuje pocet klicu, ktery uz jsme si sami spocitali
    for (int i = 0; i < pocetKlicu; i++) {
      klice[i] = st.nextToken();
    } 
  }
  
  /**
   * Urci a vrati nejmensi pocet otoceni vedouci k otevreni trezoru pomoci postupneho pridavani vzestupne serazenych hran podle ohodnoceni do grafu bez hran. Hrana je pridana pouze, pokud mezi jejimi vrcholy v grafu zatim neexistuje cesta (zjistovano pomoci DFS)
   * @return nejmensi pocet otoceni vedouci k otevreni trezoru
   */        
  public int urciNejmensiPocetOtoceni() {
    naplnPrechodyMeziKlici();
    seradHrany();
    
    for (int i = 0; i < prechodyMeziKlici.length; i++) { //Smazat vsechny prechody mezi klici (tj. hrany)
      prechodyMeziKlici[i][i] = BILA;
            
      for (int j = i + 1; j < prechodyMeziKlici[i].length; j++) {
        prechodyMeziKlici[i][j] = BILA;
        prechodyMeziKlici[j][i] = BILA;
      }
    }
    
    Hrana hrana = null;
    int predchoziPocetDosazitelnychKlicu = 0;
    int pocetDosazitelnychKlicu = 0;
    int soucetPouzitychHran = 0;
    
    for (int i = 0; i < pocetHran; i++) { //Postupne doplnovat hrany, otestovat, zda uz neexistuje cesta mezi vrcholy doplnovane hrany (pokud existuje, hranu nedoplnime)
      hrana = hrany[i];
      
      pocetDosazitelnychKlicu = existujeCesta(hrana.index1, hrana.index2); 
      if (!cestaExistuje) {
        prechodyMeziKlici[hrana.index1][hrana.index2] = hrana.ohodnoceni;
        prechodyMeziKlici[hrana.index2][hrana.index1] = hrana.ohodnoceni;
        soucetPouzitychHran += hrana.ohodnoceni;
      }
      
      if (pocetDosazitelnychKlicu >= pocetKlicu) {  //Pokud jsme pri urcovani cesty narazili na vsechny klice (tj. vrcholy grafu), muzeme skoncit
        break;
      }
    }
    
    int nejmensiPocetOtoceni = MAXIMALNI_POCET_OTOCENI;
    int pocetOtoceni = 0;
    String pocatek = "0000";
    for (int i = 0; i < klice.length; i++) {  //Urceni nejmensiho poctu otoceni od pocatecniho nastaveni trezoru (0000) k jednomu z klicu
      pocetOtoceni = urciNejmensiPocetOtoceniMeziKlici(pocatek, klice[i]);
      if (pocetOtoceni < nejmensiPocetOtoceni) {
        nejmensiPocetOtoceni = pocetOtoceni;
      }
    }
    //System.out.println("Nejmensi pocet otoceni" + nejmensiPocetOtoceni);
    
    return soucetPouzitychHran + nejmensiPocetOtoceni;
  }
  
  /**
   * Naplni prechody mezi klici, tj. naplni graf klicu ohodnocenymi hranami prechodu mezi nimy (kazdy vrchol je propojen s kazdym)
   */     
  private void naplnPrechodyMeziKlici() {
    prechodyMeziKlici = new int[pocetKlicu][pocetKlicu];
    
    pocetHran = ((pocetKlicu - 1) * pocetKlicu) / 2;
    hrany = new Hrana[pocetHran];
    
    int indexHrany = 0;
    for (int i = 0; i < prechodyMeziKlici.length; i++) {
      prechodyMeziKlici[i][i] = BILA;
            
      for (int j = i + 1; j < prechodyMeziKlici[i].length; j++) {
        prechodyMeziKlici[i][j] = urciNejmensiPocetOtoceniMeziKlici(klice[i], klice[j]);
        prechodyMeziKlici[j][i] = prechodyMeziKlici[i][j];
        hrany[indexHrany] = new Hrana(prechodyMeziKlici[i][j], i, j);
        indexHrany++;
      }
    }
    
    /*for (int i = 0; i < prechodyMeziKlici.length; i++) {
      System.out.println(java.util.Arrays.toString(prechodyMeziKlici[i]));
    }
    System.out.println("Hrany " + java.util.Arrays.toString(hrany));
    */
  }
  
  /**
   * Seradi hrany vzestupne podle ohodnoceni
   */        
  private void seradHrany() {
    Hrana pomocna = null;
    
    /*java.util.Arrays.sort(hrany, new java.util.Comparator<Hrana>() {
      public int compare(Hrana h1, Hrana h2) {
        if (h1.ohodnoceni < h2.ohodnoceni) {
          return -1;
        }
        else if (h1.ohodnoceni > h2.ohodnoceni) {
          return 1;
        }
        else {
          return 0;
        }
      }
    });*/
    java.util.Arrays.sort(hrany);
    /*for (int i = 0; i < pocetHran - 1; i++) {
      for (int j = i; j < pocetHran; j++) {
        if (hrany[i].ohodnoceni > hrany[j].ohodnoceni) {
          pomocna = hrany[i];
          hrany[i] = hrany[j];
          hrany[j] = pomocna;
        }
      }
    }*/
    //System.out.println("Serazene hrany " + java.util.Arrays.toString(hrany));
  }
  
  /**
   * Urci, zda existuje cesta mezi prvnim a druhym vrcholem pomoci DFS a ulozi vysledek do promenne <code>cestaExistuje</code>. Vrati pocet navstivenych vrcholu.
   * @param vrchol1 prvni vrchol
   * @param vrchol2 druhy vrchol
   * @return pocet navstivenych vrcholu     
   */                   
  private int existujeCesta(int vrchol1, int vrchol2) {
    int pocetNavstivenychVrcholu = 0;
    
    for (int i = 0; i < prechodyMeziKlici.length; i++) {
      prechodyMeziKlici[i][i] = BILA;
    }
    cestaExistuje = false;
    pocetNavstivenychVrcholu = dfs(vrchol1, vrchol2);
    
    return pocetNavstivenychVrcholu;
  }
  
  /**
   * Provede prohledani do hloubky z prvniho do druheho vrcholu. Skocni, kdyz se dostane do druheho vrcholu, nebo kdyz projde vsechny vrcholy. Vrati pocet nasvtivenych vrcholu
   * @param vrchol1 prvni vrchol
   * @param vrchol2 druhy vrchol
   * @return pocet nasvtivenych vrcholu      
   */        
  private int dfs(int vrchol1, int vrchol2) {
    if (cestaExistuje) {
      return 1;
    }
    
    prechodyMeziKlici[vrchol1][vrchol1] = SEDA;
    int pocetNavstivenychVrcholu = 1;
    
    
    if (vrchol1 == vrchol2) {
      cestaExistuje = true;
      return pocetNavstivenychVrcholu; 
    }
    
    for (int i = 0; i < prechodyMeziKlici.length; i++) {
      if (prechodyMeziKlici[vrchol1][i] > BILA && prechodyMeziKlici[i][i] == BILA) {
        pocetNavstivenychVrcholu += dfs(i, vrchol2);
      }
    }
    prechodyMeziKlici[vrchol1][vrchol1] = CERNA;
    
    return pocetNavstivenychVrcholu;
  }
      
  /**
   * Urci a vrati nejmensi mozny pocet otoceni, kterym se z prvniho klice stane druhy klic
   * @param klic1 prvni klic                                                                
   * @param klic2 druhy klic
   * @return nejmensi mozny pocet otoceni, kterym se z prvniho klice stane druhy klic   
   */           
  private int urciNejmensiPocetOtoceniMeziKlici(String klic1, String klic2) {
    int celkovyPocetOtoceni = 0;
    int pocetOtoceniCislice = 0;
    
    for (int i = 0; i < DELKA_KLICE; i++) {
      pocetOtoceniCislice = Math.abs((klic1.charAt(i) - '0') - (klic2.charAt(i) - '0'));
      
      if (pocetOtoceniCislice > 5) {
        pocetOtoceniCislice = 10 - pocetOtoceniCislice;
      }
      
      celkovyPocetOtoceni += pocetOtoceniCislice;
    }
    
    return celkovyPocetOtoceni;
  }
}

/**
 * Trida reprezentujici hranu grafu
 * @author Tomas Potuzak
 */  
class Hrana implements Comparable<Hrana> {
  /** Ohodnoceni hrany */
  public int ohodnoceni;
  /** Index prvniho vrcholu */
  public int index1;
  /** Index druheho vrcholu */
  public int index2;
  
  /**
   * Vytvori novou instanci hrany
   * @param ohodnoceni ohodnoceni hrany
   * @param index1 index prvniho vrcholu
   * @param index2 index druheho vrcholu  
   */           
  public Hrana(int ohodnoceni, int index1, int index2) {
    this.ohodnoceni = ohodnoceni;
    this.index1 = index1;
    this.index2 = index2;
  }
  
  @Override
  public int compareTo(Hrana h) {
    if (ohodnoceni < h.ohodnoceni) {
      return -1;
    }
    else if (ohodnoceni > h.ohodnoceni) {
      return 1;
    }
    else {
      return 0;
    }
  }
  
  /*@Override
  public String toString() {
    return "(" + ohodnoceni + ", " + index1 + " <-> " + index2 + ")"; 
  }*/
}

File: Zadání/pilsprog/1525-trezor/Trezor.pas
program Trezor;

const
  (* Barva vrcholu pro DFS - bila *)
  BILA = -1;
  (* Barva vrcholu pro DFS - seda *)
  SEDA = -2;
  (* Barva vrcholu pro DFS - cerna *)
  CERNA = -3;
  (* Delka klice k trezoru (pocet cislic) *)
  DELKA_KLICE = 4;
  (* Maximalni mozny pocet otoceni *)
  MAXIMALNI_POCET_OTOCENI = 10000;
  
type
  TStringTokenizer = record
    stringToSplit: ansistring;
    splitCharacter: char;
    remainingString: ansistring;
  end;
  THrana = record
    ohodnoceni: longint;
    index1: longint;
    index2: longint;
  end;
  TKlice = array of ansistring;
  TPrechodyMeziKlici = array of array of longint;
  THrany = array of THrana;
  TJedenTrezor = record
    (* Vsechny klice k trezoru *)
    klice: TKlice;
    (* Prechody mezi klici - matice (neorientovanych) hran grafu, na hlavni diagonale je ulozena barva vrcholu, mimo hlavni diagonalu hodnata BILA znamena, ze tam neni hrana *)
    prechodyMeziKlici: TPrechodyMeziKlici;
    (* Pocet vsech klicu *)
    pocetKlicu: longint;
    (* Pocet vsech neorientovanych hran (prechodu mezi klici) *)
    pocetHran: longint;
    (* Pole vsech hran (prechodu mezi klici) *)
    hrany: THrany;
    (* Promenna, do ktere DFS zaznamena, ze naslo cestu mezi dvema vrcholy *)
    cestaExistuje: boolean;
  end;

var
  pocetTrezoru: longint;
  radka: ansistring;
  jedenTrezor: TJedenTrezor;
  i: longint;

(*
 *
 * @param this
 * @param stringToSplit  
 * @param splitCharacter 
 *)
procedure TStringTokenizer_init(var this: TStringTokenizer; stringToSplit: ansistring; splitCharacter: char);
begin
  this.stringToSplit := stringToSplit;
  this.remainingString := stringToSplit;
  this.splitCharacter := splitCharacter;
end;

(*
 * 
 * @param this
 * @return
 *)
function TStringTokenizer_hasNextToken(var this: TStringTokenizer): boolean;
begin
  TStringTokenizer_hasNextToken := (length(this.remainingString) = 0);
end;

(*
 *
 * @param this
 * @return 
 *)
function TStringTokenizer_nextToken(var this: TStringTokenizer): ansistring;
var
  nextToken: ansistring;
  splitCharacterIndex: integer;
begin
  splitCharacterIndex := pos(this.splitCharacter, this.remainingString);
  //writeln('Nalezeno: ', splitCharacterIndex);
  if (splitCharacterIndex > 0) then begin
    nextToken := copy(this.remainingString, 1, splitCharacterIndex - 1);
    if (splitCharacterIndex < length(this.remainingString)) then begin
      this.remainingString := copy(this.remainingString, splitCharacterIndex + 1, length(this.remainingString) - splitCharacterIndex);  
    end
    else begin
      this.remainingString :='';
    end;
    
    TStringTokenizer_nextToken := nextToken;
  end
  else if (length(this.remainingString) > 0) then begin
    nextToken := this.remainingString;
    this.remainingString := '';
    TStringTokenizer_nextToken := nextToken;
  end                           
  else begin            
    TStringTokenizer_nextToken := '';                     
  end;
end;

(*
 * 
 * @param this
 * @return 
 *)
function TStringTokenizer_countTokens(var this: TStringTokenizer): longint;
var
  splitCharacterCount: longint;
  i: longint;
begin
  splitCharacterCount := 0;
  for i := 1 to length(this.stringToSplit) do begin
    if (this.stringToSplit[i] = this.splitCharacter) then begin
      inc(splitCharacterCount);
    end;
  end;
  
  TStringTokenizer_countTokens := splitCharacterCount + 1;
end;

(*
 * Vytvori novou instanci hrany
 * @param ohodnoceni ohodnoceni hrany
 * @param index1 index prvniho vrcholu
 * @param index2 index druheho vrcholu  
 *)           
procedure THrana_init(var this: THrana; ohodnoceni: longint; index1: longint; index2: longint);
begin  
  this.ohodnoceni := ohodnoceni;
  this.index1 := index1;
  this.index2 := index2;
end;

function THrana_compareTo(var this: THrana; h: THrana): longint;
begin
  if (this.ohodnoceni < h.ohodnoceni) then begin
    THrana_compareTo := -1;
  end
  else if (this.ohodnoceni > h.ohodnoceni) then begin
    THrana_compareTo := 1;
  end
  else begin
    THrana_compareTo := 0;
  end;
end;

(*
 * Vytvori novou instanci trezoru ze zadaneho retezce klicu 
 * @param trezorStr retezec obsahujici vsechny klice (s poctem klicu jako prvni cislo)
 *)        
procedure TJedenTrezor_init(var this: TJedenTrezor; trezorStr: ansistring);
var
  st: TStringTokenizer;
  i: longint;
  pomocna: ansistring;
begin
  TStringTokenizer_init(st, trezorStr, ' ');
  this.pocetKlicu := TStringTokenizer_countTokens(st) - 1;
  
  setLength(this.klice, this.pocetKlicu);
  pomocna := TStringTokenizer_nextToken(st); //Zahozeni prvniho tokenu, obsahuje pocet klicu, ktery uz jsme si sami spocitali
  for i := 0 to (this.pocetKlicu - 1) do begin
    this.klice[i] := TStringTokenizer_nextToken(st);
  end; 
end;

(*
 * Urci a vrati nejmensi mozny pocet otoceni, kterym se z prvniho klice stane druhy klic
 * @param klic1 prvni klic                                                                
 * @param klic2 druhy klic
 * @return nejmensi mozny pocet otoceni, kterym se z prvniho klice stane druhy klic   
 *)           
function TJedenTrezor_urciNejmensiPocetOtoceniMeziKlici(var this: TJedenTrezor;klic1: ansistring; klic2: ansistring): longint;
var
  celkovyPocetOtoceni: longint;
  pocetOtoceniCislice: longint;
  i: longint;

begin
  celkovyPocetOtoceni := 0;
  pocetOtoceniCislice := 0;
  
  for i := 1 to DELKA_KLICE do begin
    pocetOtoceniCislice := abs((ord(klic1[i]) - ord('0')) - (ord(klic2[i]) - ord('0')));
    
    if (pocetOtoceniCislice > 5) then begin
      pocetOtoceniCislice := 10 - pocetOtoceniCislice;
    end;
    
    celkovyPocetOtoceni := celkovyPocetOtoceni + pocetOtoceniCislice;
  end;
  
  TJedenTrezor_urciNejmensiPocetOtoceniMeziKlici := celkovyPocetOtoceni;
end;

(*
 * Naplni prechody mezi klici, tj. naplni graf klicu ohodnocenymi hranami prechodu mezi nimy (kazdy vrchol je propojen s kazdym)
 *)     
procedure TJedenTrezor_naplnPrechodyMeziKlici(var this: TJedenTrezor);
var
  indexHrany: longint;
  i, j: longint;

begin
  setLength(this.prechodyMeziKlici, this.pocetKlicu, this.pocetKlicu);
  this.pocetHran := ((this.pocetKlicu - 1) * this.pocetKlicu) div 2;
  setLength(this.hrany, this.pocetHran);
  
  indexHrany := 0;
  for i := 0 to (length(this.prechodyMeziKlici) - 1) do begin
    this.prechodyMeziKlici[i, i] := BILA;
          
    for j := i + 1 to (length(this.prechodyMeziKlici[i]) - 1) do begin
      this.prechodyMeziKlici[i, j] := TJedenTrezor_urciNejmensiPocetOtoceniMeziKlici(this, this.klice[i], this.klice[j]);
      this.prechodyMeziKlici[j, i] := this.prechodyMeziKlici[i, j];
      THrana_init(this.hrany[indexHrany], this.prechodyMeziKlici[i, j], i, j);
      inc(indexHrany);
    end;
  end;
end;


procedure TJedenTrezor_opravVrchol(var this: TJedenTrezor; spodek: longint; vrchniIndex: longint; razeni: longint);  
var
  tmp: THrana;
  naslednik: longint;

begin  
  tmp := this.hrany[vrchniIndex]; 
  naslednik := vrchniIndex * 2 + 1; 
  
  if ((naslednik < spodek) AND (THrana_compareTo(this.hrany[naslednik], this.hrany[naslednik + 1]) = razeni)) then begin 
    naslednik := naslednik + 1; 
  end;
   
  while ((naslednik <= spodek) AND (THrana_compareTo(tmp, this.hrany[naslednik]) = razeni)) do begin 
    this.hrany[vrchniIndex] := this.hrany[naslednik]; 
    vrchniIndex := naslednik; 
    naslednik := naslednik * 2 + 1; 
    
    if ((naslednik < spodek) AND (THrana_compareTo(this.hrany[naslednik], this.hrany[naslednik + 1]) = razeni)) then begin 
      naslednik := naslednik + 1; 
    end; 
  end; 
  
  this.hrany[vrchniIndex] := tmp; 
end; 
  
procedure TJedenTrezor_prohod(var this: TJedenTrezor; leva: longint; prava: longint);
var
  tmp: THrana;
  
begin 
  tmp := this.hrany[prava];     
  this.hrany[prava] := this.hrany[leva]; 
  this.hrany[leva] := tmp; 
end;

procedure TJedenTrezor_seradHrany(var this: TJedenTrezor; sestupne: boolean);
var
  i: longint;
  razeni: integer;
  
begin
  if (sestupne) then begin
    razeni := 1;
  end
  else begin
    razeni := -1;
  end;
   
  for i := (length(this.hrany) DIV 2 - 1) downto 0 do begin 
    TJedenTrezor_opravVrchol(this, length(this.hrany) - 1, i, razeni); 
  end; 

  for i := (length(this.hrany) - 1) downto 1 do begin 
    TJedenTrezor_prohod(this, 0, i); 
    TJedenTrezor_opravVrchol(this, i - 1, 0, razeni); 
  end; 
end;

(*
 * Provede prohledani do hloubky z prvniho do druheho vrcholu. Skocni, kdyz se dostane do druheho vrcholu, nebo kdyz projde vsechny vrcholy. Vrati pocet nasvtivenych vrcholu
 * @param vrchol1 prvni vrchol
 * @param vrchol2 druhy vrchol
 * @return pocet nasvtivenych vrcholu      
 *)        
function TJedenTrezor_dfs(var this: TJedenTrezor; vrchol1: longint; vrchol2: longint): longint;
var
  pocetNavstivenychVrcholu: longint;
  i: longint;
  
begin  
  if (this.cestaExistuje) then begin
    TJedenTrezor_dfs := 1;
  end;
  
  this.prechodyMeziKlici[vrchol1, vrchol1] := SEDA;
  pocetNavstivenychVrcholu := 1;
  
  
  if (vrchol1 = vrchol2) then begin
    this.cestaExistuje := true;
    TJedenTrezor_dfs := pocetNavstivenychVrcholu; 
  end;
  
  for i := 0 to (length(this.prechodyMeziKlici) - 1) do begin
    if ((this.prechodyMeziKlici[vrchol1, i] > BILA) AND (this.prechodyMeziKlici[i, i] = BILA)) then begin
      pocetNavstivenychVrcholu := pocetNavstivenychVrcholu + TJedenTrezor_dfs(this, i, vrchol2);
    end;
  end;
  this.prechodyMeziKlici[vrchol1, vrchol1] := CERNA;
  
  TJedenTrezor_dfs := pocetNavstivenychVrcholu;
end;

(*
 * Urci, zda existuje cesta mezi prvnim a druhym vrcholem pomoci DFS a ulozi vysledek do promenne <code>cestaExistuje</code>. Vrati pocet navstivenych vrcholu.
 * @param vrchol1 prvni vrchol
 * @param vrchol2 druhy vrchol
 * @return pocet navstivenych vrcholu     
 *)                   
function TJedenTrezor_existujeCesta(var this: TJedenTrezor; vrchol1: longint; vrchol2: longint): longint;
var
  pocetNavstivenychVrcholu: longint;
  i: longint;
begin  
  pocetNavstivenychVrcholu := 0;
  
  for i := 0 to (length(this.prechodyMeziKlici) - 1) do begin
    this.prechodyMeziKlici[i, i] := BILA;
  end;
  this.cestaExistuje := false;
  pocetNavstivenychVrcholu := TJedenTrezor_dfs(this, vrchol1, vrchol2);
  
  TJedenTrezor_existujeCesta := pocetNavstivenychVrcholu;
end;

(*
 * Urci a vrati nejmensi pocet otoceni vedouci k otevreni trezoru pomoci postupneho pridavani vzestupne serazenych hran podle ohodnoceni do grafu bez hran. Hrana je pridana pouze, pokud mezi jejimi vrcholy v grafu zatim neexistuje cesta (zjistovano pomoci DFS)
 * @return nejmensi pocet otoceni vedouci k otevreni trezoru
 *)        
function TJedenTrezor_urciNejmensiPocetOtoceni(var this: TJedenTrezor): longint;
var
  hrana: THrana;
  predchoziPocetDosazitelnychKlicu: longint;
  pocetDosazitelnychKlicu: longint;
  soucetPouzitychHran: longint;
  nejmensiPocetOtoceni: longint;
  pocetOtoceni: longint;
  pocatek: ansistring;
  i, j: longint;

begin  
  TJedenTrezor_naplnPrechodyMeziKlici(this);
  TJedenTrezor_seradHrany(this, false);
  
  for i := 0 to (length(this.prechodyMeziKlici) - 1) do begin //Smazat vsechny prechody mezi klici (tj. hrany)
    this.prechodyMeziKlici[i, i] := BILA;
          
    for j := i + 1 to (length(this.prechodyMeziKlici[i]) - 1) do begin
      this.prechodyMeziKlici[i, j] := BILA;
      this.prechodyMeziKlici[j, i] := BILA;
    end;
  end;
  
  predchoziPocetDosazitelnychKlicu := 0;
  pocetDosazitelnychKlicu := 0;
  soucetPouzitychHran := 0;
  
  for i := 0 to (this.pocetHran - 1) do begin //Postupne doplnovat hrany, otestovat, zda uz neexistuje cesta mezi vrcholy doplnovane hrany (pokud existuje, hranu nedoplnime)
    hrana := this.hrany[i];
    
    pocetDosazitelnychKlicu := TJedenTrezor_existujeCesta(this, hrana.index1, hrana.index2); 
    if (NOT this.cestaExistuje) then begin
      this.prechodyMeziKlici[hrana.index1, hrana.index2] := hrana.ohodnoceni;
      this.prechodyMeziKlici[hrana.index2, hrana.index1] := hrana.ohodnoceni;
      soucetPouzitychHran := soucetPouzitychHran + hrana.ohodnoceni;
    end;
    
    if (pocetDosazitelnychKlicu >= this.pocetKlicu) then begin  //Pokud jsme pri urcovani cesty narazili na vsechny klice (tj. vrcholy grafu), muzeme skoncit
      break;
    end;
  end;
  
  nejmensiPocetOtoceni := MAXIMALNI_POCET_OTOCENI;
  pocetOtoceni := 0;
  pocatek := '0000';
  for i := 0 to (length(this.klice) - 1) do begin  //Urceni nejmensiho poctu otoceni od pocatecniho nastaveni trezoru (0000) k jednomu z klicu
    pocetOtoceni := TJedenTrezor_urciNejmensiPocetOtoceniMeziKlici(this, pocatek, this.klice[i]);
    if (pocetOtoceni < nejmensiPocetOtoceni) then begin
      nejmensiPocetOtoceni := pocetOtoceni;
    end;
  end;
  
  TJedenTrezor_urciNejmensiPocetOtoceni := soucetPouzitychHran + nejmensiPocetOtoceni;
end;
  
begin
  readln(pocetTrezoru);
       
  for i := 0 to (pocetTrezoru - 1) do begin
    readln(radka);
    TJedenTrezor_init(jedenTrezor, radka);
    writeln(TJedenTrezor_urciNejmensiPocetOtoceni(jedenTrezor));
  end;

end.